### PR TITLE
Adds support for Folding Pipes in Java

### PR DESCRIPTION
**Pull request by oldergod**

### CODE REVIEW

File: okio/src/main/java/okio/Pipe.java [hunk 1]
```diff
@@ -56,43 +60,128 @@ public final Sink sink() {
...
           }
+        }
+      }
 
-          long bytesToWrite = Math.min(bufferSpaceAvailable, byteCount);
-          buffer.write(source, bytesToWrite);
-          byteCount -= bytesToWrite;
-          buffer.notifyAll(); // Notify the source that it can resume reading.
+      if (delegate != null) {
+        timeoutFolder.push(this.timeout());
```
Comment (oldergod) [inline okio/src/main/java/okio/Pipe.java]: Regard to `timeoutFolder`, should I do something similar to what we do for getting `folderSink` out of the synchronized block?

File: okio/src/main/java/okio/Pipe.java [hunk 2]
```diff
@@ -56,43 +60,128 @@ public final Sink sink() {
...
-        while (byteCount > 0) {
-          if (sourceClosed) throw new IOException("source is closed");
-
-          long bufferSpaceAvailable = maxBufferSize - buffer.size();
-          if (bufferSpaceAvailable == 0) {
-            timeout.waitUntilNotified(buffer); // Wait until the source drains the buffer.
-            continue;
+        while (byteCount > 0 && delegate == null) {
+          if (foldedSink != null) {
+            delegate = foldedSink;
```
Comment (swankjesse) [inline okio/src/main/java/okio/Pipe.java]: `break;` so you don’t need an else?

Comment (swankjesse) [reply okio/src/main/java/okio/Pipe.java]: ... and then you also don’t need the delegate check on the loop’s condition

File: okio/src/main/java/okio/Pipe.java (hunk 1 above)
Comment (swankjesse) [reply okio/src/main/java/okio/Pipe.java]: yes please. It’s the easiest to reason about

File: okio/src/test/java/okio/PipeTest.java [hunk 3]
```diff
@@ -357,6 +372,195 @@
...
+
+    final BufferedSink pipeSink = Okio.buffer(pipe.sink());
+    pipeSink.writeUtf8("hello");
+    pipeSink.emit();
+
+    final BufferedSource pipeSource = Okio.buffer(pipe.source());
+    assertEquals("hello", pipeSource.readUtf8(5));
+
+    final Buffer foldedSinkBuffer = new Buffer();
+    final boolean[] foldedSinkClosed = {false};
```
Comment (swankjesse) [inline okio/src/test/java/okio/PipeTest.java]: For what it’s worth, AtomicBoolean is much nicer for this job.

**Comment by swankjesse** [comment]
Merged to land this! Travis is unhappy but I verified manually.

```

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Okio (Parent) ...................................... SUCCESS [  0.920 s]
[INFO] Okio ............................................... SUCCESS [02:20 min]
[INFO] Okio Performance Benchmarks ........................ SUCCESS [  2.539 s]
[INFO] Okio Samples ....................................... SUCCESS [  0.410 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

File: okio/src/test/java/okio/PipeTest.java (hunk 3 above)
Comment (oldergod) [reply okio/src/test/java/okio/PipeTest.java]: Thank you! The IDE suggested I turned to bool into an array and I didn't think twice

**Pull request merged**